src/game/common/beziersegment.cpp

```diff
@@ -0,0 +1,186 @@
+ *            2 of the License, or (at your option) any later version.
+ *            A full copy of the GNU General Public License can be found in
+ *            LICENSE
+ */
+#include "beziersegment.h"
+#include "bezfwditerator.h"
+
+const D3DXMATRIX BezierSegment::s_bezBasisMatrix =
+    D3DXMATRIX(-1.0f, 3.0f, -3.0f, 1.0f, 3.0f, -6.0f, 3.0f, 0.0f, -3.0f, 3.0f, 0.0f, 0.0f, 1.0f, 0.0f, 0.0f, 0.0f);
+
+namespace
+{
+float Dot_Product(const D3DXVECTOR4 &a, const D3DXVECTOR4 &b)
+{
+    return a.x * b.x + a.y * b.y + a.z * b.z + a.w * b.w;
+}
+} // namespace
+
+BezierSegment::BezierSegment()
+{
+    for (Coord3D &point : m_points) {
+        point.Zero();
+    }
+}
+
+BezierSegment::BezierSegment(float p0x,
+    float p0y,
+    float p0z,
+    float p1x,
+    float p1y,
+    float p1z,
+    float p2x,
+    float p2y,
+    float p2z,
+    float p3x,
+    float p3y,
+    float p3z)
+{
+    m_points[0].x = p0x;
+    m_points[0].y = p0y;
+    m_points[0].z = p0z;
+
+    m_points[1].x = p1x;
+    m_points[1].y = p1y;
+    m_points[1].z = p1z;
+
+    m_points[2].x = p2x;
+    m_points[2].y = p2y;
+    m_points[2].z = p2z;
+
+    m_points[3].x = p3x;
+    m_points[3].y = p3y;
+    m_points[3].z = p3z;
+}
+
+BezierSegment::BezierSegment(float xyz_points[12])
+{
+    m_points[0].x = xyz_points[0];
+    m_points[0].y = xyz_points[1];
+    m_points[0].z = xyz_points[2];
+
+    m_points[1].x = xyz_points[3];
+    m_points[1].y = xyz_points[4];
+    m_points[1].z = xyz_points[5];
+
+    m_points[2].x = xyz_points[6];
+    m_points[2].y = xyz_points[7];
+    m_points[2].z = xyz_points[8];
+
+    m_points[3].x = xyz_points[9];
+    m_points[3].y = xyz_points[10];
+    m_points[3].z = xyz_points[11];
+}
+
+BezierSegment::BezierSegment(const Coord3D &p0, const Coord3D &p1, const Coord3D &p2, const Coord3D &p3)
+{
+    m_points[0] = p0;
+    m_points[1] = p1;
+    m_points[2] = p2;
+    m_points[3] = p3;
+}
+
+BezierSegment::BezierSegment(const Coord3D points[4])
+{
+    m_points[0] = points[0];
+    m_points[1] = points[1];
+    m_points[2] = points[2];
+    m_points[3] = points[3];
+}
+
+void BezierSegment::Evaluate_Bez_Segment_At_T(float t, Coord3D *point)
+{
+    if (point != nullptr) {
+        float sqr = t * t;
+        float cub = t * t * t;
+
+        D3DXVECTOR4 vecW(cub, sqr, t, 1.0f);
+        D3DXVECTOR4 vecX(m_points[0].x, m_points[1].x, m_points[2].x, m_points[3].x);
+        D3DXVECTOR4 vecY(m_points[0].y, m_points[1].y, m_points[2].y, m_points[3].y);
+        D3DXVECTOR4 vecZ(m_points[0].z, m_points[1].z, m_points[2].z, m_points[3].z);
+
+        D3DXVECTOR4 vecF;
+        D3DXVec4Transform(&vecF, &vecW, &BezierSegment::s_bezBasisMatrix);
+
+        point->x = Dot_Product(vecX, vecF);
+        point->y = Dot_Product(vecY, vecF);
+        point->z = Dot_Product(vecZ, vecF);
+    }
+}
+
+void BezierSegment::Get_Segment_Points(int count, std::vector<Coord3D> *points)
+{
+    if (points != nullptr) {
+        points->resize(count);
```

Comment:
There appears to be a `points->clear()` before too, but its useless, so omitted.